File: 515/FindLargestValueInEachTreeRow.cpp
/*
https://leetcode.com/problems/find-largest-value-in-each-tree-row/description/?envType=daily-question&envId=2024-12-25
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    vector<int> bfs(TreeNode *root){
        vector<int>res;
        queue<TreeNode*>q;
        q.push(root);
        while(!q.empty()){
            int sz=q.size(), curmax=INT_MIN;
            while(sz--){
                auto curr=q.front();
                q.pop();
                curmax=max(curmax, curr->val);
                if(curr->left!=nullptr){
                    q.push(curr->left);
                }
                if(curr->right!=nullptr){
                    q.push(curr->right);
                }
            }
            res.push_back(curmax);
        }
        return res;
    }

    vector<int> largestValues(TreeNode* root) {
        if(root==nullptr) return {};
        return bfs(root);
    }
};

File: 515/FindLargestValueInEachTreeRow.java
/*
https://leetcode.com/problems/find-largest-value-in-each-tree-row/description/?envType=daily-question&envId=2024-12-25
*/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    public static List<Integer> breadthFirstSearch_forTrees(TreeNode root){
        List<Integer>resultantList=new ArrayList<>();
        LinkedList<TreeNode>queueOfTreeNodes=new LinkedList<>();
        queueOfTreeNodes.add(root);
        while(!queueOfTreeNodes.isEmpty()){
            int sizeOfQueue=queueOfTreeNodes.size(), currentMaxInRow=Integer.MIN_VALUE;
            while(sizeOfQueue-- >0){
                TreeNode currentNode=queueOfTreeNodes.pollFirst();
                int currentNodeValue=currentNode.val;
                currentMaxInRow=Math.max(currentMaxInRow, currentNodeValue);
                if(currentNode.left!=null){
                    queueOfTreeNodes.add(currentNode.left);
                }
                if(currentNode.right!=null){
                    queueOfTreeNodes.add(currentNode.right);
                }
            }
            resultantList.add(currentMaxInRow);
        }
        return resultantList;
    }

    public List<Integer> largestValues(TreeNode root) {
        if(root==null) return new ArrayList<>();
        return breadthFirstSearch_forTrees(root);
    }
}